pkg/policy/mapstate.go

```diff
@@ -128,38 +128,22 @@ func (e *MapStateEntry) DatapathEqual(o *MapStateEntry) bool {
 func (keys MapState) DenyPreferredInsert(key Key, entry MapStateEntry) {
 	// If it is a deny policy, it will take precedence over any other key.
 	if entry.IsDeny {
-		// If there is an allow or a deny L3-L4 and the new deny is a L3-Only,
-		// the new deny takes precedence over all L3-L4 for each the L3 is the
-		// same on both the existing rules and new deny rule.
+		// If we are adding a "deny" L3-only, we need to also add a deny L3-L4
+		// for all the allow L4-only cases. This is necessary because the
+		// datapath checks for a presence of a L3-L4, followed by a L4-only,
+		// followed by a L3-only.
 		if key.DestPort == 0 {
-			for k := range keys {
+			for k, v := range keys {
 				if k.TrafficDirection == key.TrafficDirection &&
 					k.Identity == key.Identity {
```

Comment:
This loop is only handling cases where the L3 is the same as in the L3-only case, i.e., the L3/L4 cases where the L3 is the same as in the L3-only deny rule (`key`). Especially, L4-only keys, where the L3 is zero (right?) are not handled, while it seems that for each L4-only rule, a L3/L4 deny rule would need to be added, as L4-only keys are looked up before L3-only keys?

Reply:
Also, how are we now giving precedence to existing deny keys when adding a L3/L4 or L4-only allow key?